File: Algorithms/practical-guide-part2/Datastructure/Stack.js
// Implement stack data structure

class Stack {
  constructor() {
    this._storage = {};
    this._length = 0;
  }

  push(value) {
    this._storage[this._length] = value;
    this._length++;
  }

  pop() {
    if (this._length) {
      const removedItem = this._storage[this._length - 1];
      delete this._storage[this._length];
      this._length--;
      return removedItem;
    }
  }

  peek() {
    if (this._length) {
      return this._storage[this._length - 1];
    }

    return [];
  }

  getLength() {
    return this._length;
  }
}

const myStack = new Stack();
myStack.push(10);
myStack.push(20);
myStack.push(30);
myStack.pop();
console.log(myStack);

File: Educative/DS/Stack.py
# Implement basic Stack data structure


class Stack:
    def __init__(self):
        self._storage = {}
        self._length = 0

    def push(self, data):
        if data is not None:
            self._storage[self._length] = data
            self._length += 1

            return self._length

    def pop(self):
        if self._length >= 0:
            removed_item = self._storage[self._length - 1]
            del self._storage[self._length - 1]

            self._length -= 1

            return removed_item

    def size(self):
        return self._length

    def get_stack(self):
        return self._storage

    def is_empty(self):
        return self._storage == {}

    def peek(self):
        if not self.is_empty():
            return self._storage[self._length - 1]


myStack = Stack()
myStack.push("A")
myStack.push("B")
print(myStack.get_stack())
myStack.push("C")
print(myStack.get_stack())
myStack.pop()
print(myStack.get_stack())
print(myStack.is_empty())
print(myStack.peek())
